File: iosApp/iosApp/Modules/Alert/AlertWireframe.swift
// Created by web3dgn on 02/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultAlertWireframe {
    private weak var parent: UIViewController?
    private let context: AlertWireframeContext
    
    private weak var vc: UIViewController?

    init(
        parent: UIViewController?,
        context: AlertWireframeContext
    ) {
        self.parent = parent
        self.context = context
    }
}

extension DefaultAlertWireframe {

    func present() {
        let vc = makeViewController()
        parent?.present(vc, animated: true)
    }
    
    func navigate(to destination: AlertWireframeDestination) {
        if destination is AlertWireframeDestination.Dismiss {
            vc?.navigationController?.dismiss(animated: true) ?? vc?.dismiss(animated: true)
        }
    }
}

private extension DefaultAlertWireframe {

    func makeViewController() -> UIViewController {
        let vc: AlertViewController = UIStoryboard(.alert).instantiate()
        let presenter = DefaultAlertPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            context: context
        )
        vc.presenter = presenter
        vc.contentHeight = context.contentHeight
        let nc = NavigationController(rootViewController: vc)
        nc.modalPresentationStyle = .custom
        nc.transitioningDelegate = vc
        self.vc = nc
        return nc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/alert/AlertWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.alert

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3walletcore.modules.alert.AlertWireframe
import com.sonsofcrypto.web3walletcore.modules.alert.AlertWireframeContext
import com.sonsofcrypto.web3walletcore.modules.alert.AlertWireframeDestination
import com.sonsofcrypto.web3walletcore.modules.alert.DefaultAlertPresenter

class DefaultAlertWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: AlertWireframeContext,
): AlertWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.get()?.navigationFragment?.present(fragment)
    }

    override fun navigate(destination: AlertWireframeDestination) {
        when (destination) {
            is AlertWireframeDestination.Dismiss -> {
                parent?.get()?.navigationFragment?.dismiss()
            }
        }
    }

    private fun wireUp(): Fragment {
        val fragment = AlertFragment()
        val presenter = DefaultAlertPresenter(
            WeakRef(fragment),
            this,
            context
        )
        fragment.presenter = presenter
        return fragment
    }
}